### PR TITLE
Deprecate SplitYamlProvider, subsume functionality into YamlProvider w/a refactor

### PR DESCRIPTION
**Pull request by ross**
This PR refactors/reworks YamlProvider so that it supports both the single file and split zone setups or any combination thereof. 

As part of this SplitYamlProvider is deprecated and will be removed in 2.0. The previous behavior can be achieved via YamlProvider by using `split_extension: [.ext]`, `split_only: true`, and `split_catchall: true`. 

/cc Fixes https://github.com/octodns/octodns/issues/1040 @viranch 
/cc https://github.com/octodns/octodns/issues/962 since this work was sort of blocking plans there

### CODE REVIEW

File: octodns/provider/yaml.py
Comment (ross) [reply octodns/provider/yaml.py:71]: See the `split_catchall` option. It's defaulting to `False` which means each record name gets its own file with the data for that name. OSX and Linux have no problems with files named `.yaml` and `*.yaml` and it seems cleaner to do it this way. If you use the SplitYamlProvider shim it'll get the previous/expected behavior.

Not sure about Windows, but either way the catchall behavior can be enabled by setting `split_catchall` and that file will be created with the APEX and/or wildcard data.

Should be documented in the param doc up above. I started to mention it down here as well, but the doc was getting pretty long.

**Comment by ross** [comment]
> Wow that's a first. Tests pass, but some sort of coverage issue only on 3.8 and 3.9 https://github.com/octodns/octodns/actions/runs/5893500410/job/15985179228
> 
> I can't reproduce with the 3.9 I have locally. Still digging.

Hrm. Can't reproduce this at all. Even in a linux docker container with 3.9:

```console
## environment & versions ######################################################
Python 3.9.17
pip 23.2.1 from /tmp/octodns/env/lib/python3.9/site-packages/pip (python 3.9)
## clean up ####################################################################
## begin #######################################################################
## lint ########################################################################
## formatting ##################################################################
All done! ✨ 🍰 ✨
107 files would be left unchanged.
## tests/coverage ##############################################################
============================================================================================================ test session starts =============================================================================================================
platform linux -- Python 3.9.17, pytest-7.4.0, pluggy-1.2.0
rootdir: /tmp/octodns
configfile: pyproject.toml
plugins: mock-3.11.1, network-0.0.1, cov-4.1.0
collected 264 items
...
Required test coverage of 100% reached. Total coverage: 100.00%

============================================================================================================ 264 passed in 5.95s =============================================================================================================
## complete ####################################################################
```

Running low on ideas to reproduce this.

**Comment by viranch** [comment]
Any way to get `htmlcov` from the CI?

**Comment by ross** [comment]
> Any way to get `htmlcov` from the CI?

I don't remember there being a way, but it's the 2nd thing I plan to look into. Going to quickly throw in a `pip freeze` as part of the versions info so that I can see what versions of things are installed.

File: octodns/provider/yaml.py
Comment (viranch) [reply octodns/provider/yaml.py:71]: Ah gotcha. Should `split_catchall` default to `true` then to match the previous `SplitYamlProvider` behavior? I don't have a preference either way personally.

Comment (ross) [reply octodns/provider/yaml.py:71]: The `[zone-name][ext]/$[zone-name].yaml` kind of confused me the first time I ran across it during this refactor, I haven't used split provider before, so that lead me to default to `false`. I assumed it had something to do with filename support on OSes so figured it's best to let people who are worried about that enable it provided they weren't coming from SplitYamlProvider where it was the default behavior.

I don't feel strongly about it, since I don't use it or plan to. So if I'm outvoted that's fine with me. Assuming you prefer true, @yzguy can be the tiebreaker :grin:

**Comment by ross** [comment]
> Any way to get `htmlcov` from the CI?

Looks like we should be able to create an artifact https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts. I'll dig into that in a separate PR. Seems like a good idea in general.

**Comment by ross** [comment]
> Going to quickly throw in a `pip freeze` as part of the versions info so that I can see what versions of things are installed.

There are no module version differences.

**Comment by viranch** [comment]
I could repro on py3.8.
[htmlcov.tar.gz](https://github.com/octodns/octodns/files/12372225/htmlcov.tar.gz)

**Comment by ross** [comment]
Thanks. Interesting. My only guess here is that it's actually some sort of race condition between creating the files and/or directory, but if that's the case I'm not sure how the tests are passing. Will dig in.

**Comment by ross** [comment]
> Thanks. Interesting. My only guess here is that it's actually some sort of race condition between creating the files and/or directory, but if that's the case I'm not sure how the tests are passing. Will dig in.

So that code was getting coverage through testing of other functionality, probably in Manager and afaict there was nothing race-y about it.

That did lead me to write more explicit testing of YamlProvider.list_zones in the provider's tests. That was pushed up in https://github.com/octodns/octodns/pull/1044/commits/608e367a9bad0ecd3236d3004da3e8b2bdef50b6. It should not hit 100% coverage from its own tests. That said it's still failing on the same 2 versions with the same coverage numbers so I'm guessing something else is going on here. Perhaps some sort of funky bug with coverage???

@viranch what OS/env were you able to reproduce on?

**Comment by ross** [comment]
diffs between versions on 3.9 which is failing and 3.10 which is passing:

```patch
coho:octodns ross$ diff -u /tmp/v9.txt /tmp/v10.txt
--- /tmp/v9.txt	2023-08-17 12:19:51.000000000 -0700
+++ /tmp/v10.txt	2023-08-17 12:19:27.000000000 -0700
@@ -1,6 +1,6 @@
 ## environment & versions ######################################################
-Python 3.9.17
-pip 23.2.1 from /home/runner/work/octodns/octodns/env/lib/python3.9/site-packages/pip (python 3.9)
+Python 3.10.12
+pip 23.2.1 from /home/runner/work/octodns/octodns/env/lib/python3.10/site-packages/pip (python 3.10)
 ## modules:
 black==23.7.0
 bleach==6.0.0
@@ -56,7 +56,6 @@
 six==1.16.0
 tomli==2.0.1
 twine==4.0.2
-typing_extensions==4.7.1
 urllib3==2.0.4
 webencodings==0.5.1
 zipp==3.16.2
```

**Comment by ross** [comment]
This seems to be some sort of coverage bug. I split up the `if`s to see if it gave another/better why message and it's saying the conditions were never true. That's definitely not correct, they are true for 3.10/11 and logically walking the code when looking at the new tests I wrote it would have to be getting into those blocks or else there'd be unwated zone names being returned from list_zones.

Screen Shot 2023-08-17 at 12 58 34 PM

I guess I should probably file a bug against coverage and see what they say. Will probably do that ~tomorrow.

**Comment by ross** [comment]
So storing the values in vars and reordering the clauses allowed coverage to pass. May have been the case that the early elements of the `if` were always failing things before the later/last one could, but it's weird that 3.10/3.11 would have been OK with it. Anyway, this way is fine and coverage is happy so this should be gtg for final 👀, 💭.

**Comment by ross** [comment]
> I probably will go ahead and file something against coverage just to see what they think/say.

Filed https://github.com/nedbat/coveragepy/issues/1672

**Comment by viranch** [comment]
> @viranch what OS/env were you able to reproduce on?

I'm on Archlinux with various pythons installed using pyenv.

**Comment by ross** [comment]
> I'm on Archlinux with various pythons installed using pyenv.

https://github.com/nedbat/coveragepy/issues/1672 answered the question. Apparently there were ~bugs in older versions of python's tracing with `continue`s. Only thing that wasn't answered was why it's only happening in some envs and not others, but guessing that may have to do with some low-level bits of what the python code gets turned into. Perhaps differently for OSX than Linux (even Linux running on Docker for Mac)

This is ready to review and move forward with. I've bene adding some more thorough tests I've come across as I work on the interaction of improvements beyond this PR.

**Comment by ross** [comment]
> Perhaps differently for OSX than Linux (even Linux running on Docker for Mac)

Was driving me nuts... I was able to repro on a linux box (in docker) so it must be an OS/arch thing of some sort.

File: octodns/provider/yaml.py
Comment (yzguy) [reply octodns/provider/yaml.py:78]: So from my quick poking around (as I've never used the SplitYamlProvider either before today)

The user specified extension (`split_extension` in this case) is definitely related to #657, I assume that also applied to directories.

The `$` thing is actually pretty annoying regardless of OS because you need to escape it or the OS will think it's a variable, regardless it seems the previous behavior is it's only created whenever the record is `''` or `*`, it's not a user configurable thing.

So in that case, I'd think to preserve the current behavior of the SplitYamlProvider, whenever a user enables splitting (with `split_extension`), it should create the same catchall file if the record is `''` or `*`, if a user does not want it, they can disable it (which would be one of the new features of this change) with `split_catchall: false`

I would have to vote for `split_catchall: true` in that case, final answer :P

Comment (ross) [reply octodns/provider/yaml.py:78]: > It would almost make more sense to just have the "catchall" be [zone-name][ext]/[zone-name].yaml, and the '' and * go in there, and we don't use any characters that the OS does special things with (variables, hidden files, etc.) in the file name

Mostly agree. Not really sure where the $ thing came from, I guess it's not a legal "name" and therefore could never exist from a record `zone.name.zone.name.` is a legal fqdn and would result in `[zone-directory]/[zone.name].yaml`. Thus you'd have the name based thing in that file along with the catchalls. Why someone would want it to I have no clue, but 🤷 

> These are likely too big of behavior changes but I'm thinking the ideal set up is

My main goal is to keep `SplitYamlProvder` unchanged and `YamlProvider` 100% backwards compatible. I'm pretty flexible as far as what the "new" behaviors are.

> disable_zonefile would not be needed, whenever split_zone is enabled, the zone directory is used

My thinking was to allow the option to partially split things, some split files and some in the main zone file. 

At GitHub I likely would have done that for the large dynamic records and then put all the simple/secondary stuff into the main file. The other use-case, which I may have done there too, would be to put important stuff in its own files with CODEOWNERS requiring specific reviews. This could even be extended to have per team files. 

Granted it could also be accomplished with just split files and using whatever filenames you want. No reason both split and the primary are required. The names of the files are actually ignored. Split is really just "directory"

I'll think about removing `disable_zonefile` in favor of split_extension turning off the zone file. That seems to make sense.

I will also enable catchall by default, but I think i'll leave it with the `$` to avoid having the odd/edge case of having to merge `zone.name.zone.name.` into that file if it happened to exist.

**Comment by ross** [comment]
This is ready for final review and/or thoughts.

File: octodns/provider/yaml.py
Comment (viranch) [reply octodns/provider/yaml.py:74]: So the final decision is to go with `$zone.yaml` format? If so, should the example be updated per my suggestion at the top?

File: octodns/provider/yaml.py
Comment (viranch) [inline octodns/provider/yaml.py:242]: Does this mean that a single instance of YamlProvider can't be used to support a hybrid of non-split and split yaml files? Looks like if I've set `split_extension` to non-empty/non-null, it will skip listing non-split zones altogether.

File: octodns/provider/yaml.py
Comment (viranch) [inline octodns/provider/yaml.py:328]: Does this mean that a single instance of YamlProvider can't be used to read a setup with a hybrid of non-split and split yaml files? Looks like if I've set `split_extension` to non-empty/non-null, this will skip reading non-split zones altogether.

File: octodns/provider/yaml.py
Comment (ross) [reply octodns/provider/yaml.py:74]: Good catch. Forgot that bit when I flipped the default. Fix pushed.

File: octodns/provider/yaml.py
Comment (ross) [reply octodns/provider/yaml.py:328]: That is sort of correct. 

I was thinking it's kind of confusing to have it look for both a directory named `zone.name./` and a file named `zone.name.yaml` at the same level and it'd be preferable to just look at `zone.name./` and then a single file, with whatever name you want could live under that directory, e.g. `zone.name./zone.name.yaml`, when you don't want to split things up. 

Don't mind adding it back if you'd specifically prefer to be able mix the two and the above isn't to your liking.

Comment (viranch) [reply octodns/provider/yaml.py:329]: Yeah I personally prefer support for hybrid setup for the use-case described in https://github.com/octodns/octodns/issues/1040.

Comment (ross) [reply octodns/provider/yaml.py]: Reverted that bit. `disable_zonefile` is an option again and defaults to false so when split_extension is enabled it'll also be looking for the standard `zone.name.yaml`

File: octodns/provider/yaml.py
Comment (viranch) [reply octodns/provider/yaml.py:79]: The example also needs to show `$github.yaml` rather than just `.yaml` since we're defaulting to that behavior with `split_catchall: true`?

**Review by viranch — APPROVED** [review]
LGTM!

**Pull request merged**
